zephyr/syscall/sof_dma.c

```diff
@@ -0,0 +1,217 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2025 Intel Corporation.
+
+#include <sof/lib/dma.h>
+#include <zephyr/kernel.h>
+#include <zephyr/internal/syscall_handler.h>
+
+#ifdef CONFIG_USERSPACE
+
+static inline bool sof_dma_has_access(struct sof_dma *dma)
+{
+	/*
+	 * use the Zephyr dma.h device handle to check calling
+	 * thread has access to it
+	 */
+	return k_object_is_valid(dma->z_dev, K_OBJ_ANY);
+}
+
+static inline bool sof_dma_is_valid(struct sof_dma *dma)
+{
+	const struct dma_info *info = dma_info_get();
+	struct sof_dma *dma_ko = NULL, *d;
+
+	if (!info->num_dmas)
+		return false;
+
+	for (d = info->dma_array; d < info->dma_array + info->num_dmas; d++) {
+		if (d == dma) {
+			dma_ko = d;
+			break;
+		}
+	}
+
+	if (dma_ko && sof_dma_has_access(dma))
+		return true;
+
+	return false;
+}
+
+static inline struct sof_dma *z_vrfy_sof_dma_get(uint32_t dir, uint32_t cap,
+						 uint32_t dev, uint32_t flags)
+{
+	struct sof_dma *dma = z_impl_sof_dma_get(dir, cap, dev, flags);
+
+	if (sof_dma_has_access(dma))
+		return dma;
```

Comment:
hm, firstly it's a bit unusual I think to check after executing the kernel function? Normally you'd want to prevent the "assuming benevolence" kernel function to execute on unsafe user parameters, but maybe in this case it's justified - the user can only pass integers. But if the user doesn't have access to that, shouldn't we `put()` before returning `NULL`?

Reply:
Good points. The order is on purpose, so let me add an longer inline comment to explain. And yes, the put() was missing, will also added that.